pageserver/src/walreceiver/connection_manager.rs

```
@@ -0,0 +1,1134 @@
+                max_lsn_wal_lag,
+            );
+            loop {
+                select! {
+                    _ = cancellation.changed() => {
+                        info!("Broker subscription init cancelled, shutting down");
+                        if let Some(wal_connection) = walreceiver_state.wal_connection.take()
+                        {
+                            wal_connection.connection_task.shutdown().await;
+                        }
+                        return Ok(());
+                    },
+
+                    _ = connection_manager_loop_step(
+                        &broker_loop_prefix,
+                        &mut client,
+                        &mut walreceiver_state,
+                    ) => {},
+                }
+            }
+        }
+        .instrument(info_span!("wal_connection_manager", id = %id))
+    })
+}
+
+/// Attempts to subscribe for timeline updates, pushed by safekeepers into the broker.
+/// Based on the updates, desides whether to start, keep or stop a WAL receiver task.
+/// If etcd subscription is cancelled, exits.
+async fn connection_manager_loop_step(
+    broker_prefix: &str,
+    etcd_client: &mut Client,
+    walreceiver_state: &mut WalreceiverState,
+) {
+    let id = walreceiver_state.id;
+
+    let mut broker_subscription =
+        subscribe_for_timeline_updates(etcd_client, broker_prefix, id).await;
+    info!("Subscribed for etcd timeline changes, waiting for new etcd data");
+
+    loop {
+        select! {
+            // the order of the polls is especially important here, since the first task to complete gets selected and the others get dropped (cancelled).
+            // place more frequetly updated tasks below to ensure the "slow" tasks are also reacted to.
+            biased;
+
+            // First, check if the broker connection is intact, otherwise we need to restart the whole step
+            broker_connection_result = &mut broker_subscription.watcher_handle => {
```

Since closing watch channel and task ending don't happen atomically, there is a chance of getting out through `broker_subscription.value_updates.recv()` arm `returning None`, in which case again error is swallowed.